Handle `exitFullscreen()` when using Quick Look

etiennesegonzac opened the pull request:
#### 211b6e10b7238bbe628003074ecf6b15cf0e6b2b
<pre>
Handle `exitFullscreen()` when using Quick Look
<a href="https://bugs.webkit.org/show_bug.cgi?id=270297">https://bugs.webkit.org/show_bug.cgi?id=270297</a>
&lt;<a href="https://rdar.apple.com/120903037">rdar://120903037</a>&gt;

Reviewed by Andy Estes and Aditya Keerthi.

Use `PreviewApplication` instead or `Preview` in order to get access to
programmatic close. Wire it to the `exitFullscreen` code path.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
(WebKit::WebFullScreenManagerProxy::imageMIMEType const): Deleted.
(WebKit::WebFullScreenManagerProxy::imageBuffer const): Deleted.
(WebKit::WebFullScreenManagerProxy::isImageElement const):
Hold the image buffer with a `RefPtr` instead of a `std::optional&lt;Ref&gt;`.
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::exitFullScreen):
(WebKit::sharedQuickLookFileQueue):
(WebKit::WebFullScreenManagerProxy::prepareQuickLookImageURL const):
Add a method to generate local URLs instead of passing NSData to the
Preview.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController exitFullScreen]):
Clean-up PreviewWindowController on exit.
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
Switch to a URL-based QLItem.
(-[WKFullScreenWindowController previewWindowControllerDidClose:]):
(-[WKFullScreenWindowController previewWindowControllerDidClose]): Deleted.
Check that the instance closing is the one we care about.
(-[WKFullScreenWindowController provideDataForItem:]): Deleted.
Remove the NSData-based delegate for QLItems.

* Source/WebKit/WebKitSwift/Preview/WKSPreviewWindowController.h:
Pass the WKSPreviewWindowController instance to the delegate.
* Source/WebKit/WebKitSwift/Preview/PreviewWindowController.swift:
(PreviewWindowController.item):
(PreviewWindowController.previewSession):
(PreviewWindowController.presentWindow):
(PreviewWindowController.dismissWindow):
(PreviewWindowController.previewApp): Deleted.
Switch to `PreviewApplication`.
Implement `dismissWindow`.

Canonical link: <a href="https://commits.webkit.org/275778@main">https://commits.webkit.org/275778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2250cbabcd3bb33837c94603411c480a9f5b7f30

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42746 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21767 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45147 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45358 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38870 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25434 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19132 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35380 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43319 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18773 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36796 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16339 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16402 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/802 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38916 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38188 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46867 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17564 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14481 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42107 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19183 "Built successfully") | | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40735 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9548 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19362 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18828 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1866,13 +1869,12 @@ - (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(Compl
-            _previewWindowController = nil;
+            manager->prepareQuickLookImageURL([strongSelf = retainPtr(self), self] (URL&& url) mutable {
+                auto item = adoptNS([PAL::allocQLItemInstance() initWithURL:url]);
+                _previewWindowController = [WebKit::allocWKSPreviewWindowControllerInstance() initWithItem:item.get()];
```
Doesn't this also need `adoptNS`?

etiennesegonzac replied:
it does!

pxlcoder commented on Source/WebKit/WebKitSwift/Preview/WKSPreviewWindowController.h:
```diff
@@ -33,7 +33,7 @@ NS_ASSUME_NONNULL_BEGIN
 
 @protocol WKSPreviewWindowControllerDelegate <NSObject>
-- (void)previewWindowControllerDidClose;
+- (void)previewWindowControllerDidClose:(id)previewWindowController;
```
Can the parameter be more strongly typed?

aestes commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
```diff
@@ -37,12 +37,21 @@
 namespace WebKit {
 using namespace WebCore;
 
+#if PLATFORM(VISION) && ENABLE(QUICKLOOK_FULLSCREEN)
```
Nit: shouldn't just `ENABLE(QUICKLOOK_FULLSCREEN)` be sufficient?

pxlcoder commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
```diff
@@ -211,6 +220,30 @@ void WebFullScreenManagerProxy::exitFullScreen()
+#if PLATFORM(VISION) && ENABLE(QUICKLOOK_FULLSCREEN)
+void WebFullScreenManagerProxy::prepareQuickLookImageURL(CompletionHandler<void(URL&&)>&& completionHandler) const
+{
+    if (!m_imageBuffer.has_value())
```
`has_value()` is unnecessary – `!m_imageBuffer` is equivalent.

aestes replied:
No need for `.has_value()`.

Also, this isn't new to this PR, but I noticed that `m_imageBuffer` is a `std::optional<Ref<WebCore::SharedBuffer>>`. Why not just make it a `RefPtr` instead?

aestes commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
```diff
@@ -211,6 +220,30 @@ void WebFullScreenManagerProxy::exitFullScreen()
     m_client.exitFullScreen();
 }
 
+#if PLATFORM(VISION) && ENABLE(QUICKLOOK_FULLSCREEN)
```
Ditto

pxlcoder reviewed (approved):
Will defer the review of Swift code to @aestes.

aestes commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
```diff
@@ -211,6 +220,30 @@ void WebFullScreenManagerProxy::exitFullScreen()
+
+    auto& buffer = m_imageBuffer.value();
+    auto& mimeType = m_imageMIMEType;
+    sharedQuickLookFileQueue().dispatch([buffer, mimeType, completionHandler = WTFMove(completionHandler)]() mutable {
```
`m_imageMIMEType` is a `String`, so you need to cross-thread copy it when passing it to another queue (e.g., `[mimeType = crossThreadCopy(m_imageMIMEType), ...]`).

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1173,6 +1168,14 @@ - (void)exitFullScreen
     _fullScreenState = WebKit::WaitingToExitFullScreen;
     _exitingFullScreen = YES;
 
+#if PLATFORM(VISION) && ENABLE(QUICKLOOK_FULLSCREEN)
```
Nit: shouldn't just `ENABLE(QUICKLOOK_FULLSCREEN)` be sufficient?

aestes reviewed (approved):
r=me but please address the cross-thread copy issue before landing.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1925,21 +1927,12 @@ - (void)hideUI
-
-@implementation WKFullScreenWindowController (QLPreviewItemDataProvider)
-- (NSData *)provideDataForItem:(QLItem *)item
+- (void)previewWindowControllerDidClose:(WKSPreviewWindowController *)previewWindowController
```
Nit: for consistency with the declaration in WKSPreviewWindowController.h you should use `id` for the type of `previewWindowController`

etiennesegonzac committed:
Handle `exitFullscreen()` when using Quick Look
https://bugs.webkit.org/show_bug.cgi?id=270297
<rdar://120903037>

Reviewed by Andy Estes and Aditya Keerthi.

Use `PreviewApplication` instead or `Preview` in order to get access to
programmatic close. Wire it to the `exitFullscreen` code path.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
(WebKit::WebFullScreenManagerProxy::imageMIMEType const): Deleted.
(WebKit::WebFullScreenManagerProxy::imageBuffer const): Deleted.
(WebKit::WebFullScreenManagerProxy::isImageElement const):
Hold the image buffer with a `RefPtr` instead of a `std::optional<Ref>`.
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::exitFullScreen):
(WebKit::sharedQuickLookFileQueue):
(WebKit::WebFullScreenManagerProxy::prepareQuickLookImageURL const):
Add a method to generate local URLs instead of passing NSData to the
Preview.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController exitFullScreen]):
Clean-up PreviewWindowController on exit.
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
Switch to a URL-based QLItem.
(-[WKFullScreenWindowController previewWindowControllerDidClose:]):
(-[WKFullScreenWindowController previewWindowControllerDidClose]): Deleted.
Check that the instance closing is the one we care about.
(-[WKFullScreenWindowController provideDataForItem:]): Deleted.
Remove the NSData-based delegate for QLItems.

* Source/WebKit/WebKitSwift/Preview/WKSPreviewWindowController.h:
Pass the WKSPreviewWindowController instance to the delegate.
* Source/WebKit/WebKitSwift/Preview/PreviewWindowController.swift:
(PreviewWindowController.item):
(PreviewWindowController.previewSession):
(PreviewWindowController.presentWindow):
(PreviewWindowController.dismissWindow):
(PreviewWindowController.previewApp): Deleted.
Switch to `PreviewApplication`.
Implement `dismissWindow`.

Canonical link: https://commits.webkit.org/275778@main
